ome_zarr/axes.py

```diff
@@ -0,0 +1,110 @@
+
+from .format import CurrentFormat, Format
+
+KNOWN_AXES = {"x": "space", "y": "space", "z": "space", "c": "channel", "t": "time"}
+
+
+class Axes:
+    def __init__(
+        self,
+        axes: Union[List[str], List[Dict[str, str]]],
+        fmt: Format = CurrentFormat(),
+    ) -> None:
+
+        if axes is not None:
+            self.axes = self._axes_to_dicts(axes)
+        self.fmt = fmt
+
+    def validate(self) -> None:
+        """Raises ValueError if not valid"""
+        if self.fmt.version in ("0.1", "0.2"):
+            return
+
+        # check names (only enforced for version 0.3)
+        if self.fmt.version == "0.3":
+            self._validate_03()
+            return
+
+        self._validate_axes_types()
+
+    def to_list(self, fmt: Format) -> Union[List[str], List[Dict[str, str]]]:
+        if fmt.version == "0.3":
+            return self._get_names()
+        return self.axes
+
+    @staticmethod
+    def _axes_to_dicts(
+        axes: Union[List[str], List[Dict[str, str]]]
+    ) -> List[Dict[str, str]]:
+        """Returns a list of axis dicts with name and type"""
+        axes_dicts = []
+        for axis in axes:
+            if isinstance(axis, str):
+                axis_dict = {"name": axis}
+                if axis in KNOWN_AXES:
+                    axis_dict["type"] = KNOWN_AXES[axis]
+                axes_dicts.append(axis_dict)
+            else:
+                axes_dicts.append(axis)
+        return axes_dicts
+
+    def _validate_axes_types(self) -> None:
+        """
+        Validate the axes types according to the spec, version 0.4+
+        """
+        axes_types = [axis.get("type") for axis in self.axes]
+        known_types = list(KNOWN_AXES.values())
+        unknown_types = [atype for atype in axes_types if atype not in known_types]
+        if len(unknown_types) > 1:
+            raise ValueError(
+                "Too many unknown axes types. 1 allowed, found: %s" % unknown_types
+            )
+
+        def _last_index(item: str, item_list: List[Any]) -> int:
+            return max(loc for loc, val in enumerate(item_list) if val == item)
+
+        if "time" in axes_types and _last_index("time", axes_types) > 0:
+            raise ValueError("'time' axis must be first dimension only")
+
+        if axes_types.count("channel") > 1:
+            raise ValueError("Only 1 axis can be type 'channel'")
+
+        if "channel" in axes_types and _last_index(
+            "channel", axes_types
+        ) > axes_types.index("space"):
+            raise ValueError("'space' axes must come after 'channel'")
+
+    def _get_names(self) -> List[str]:
+        """Returns a list of axis names"""
+        axes_names = []
+        for axis in self.axes:
+            if "name" not in axis:
+                raise ValueError("Axis Dict %s has no 'name'" % axis)
+            axes_names.append(axis["name"])
+        return axes_names
+
+    def _validate_03(self) -> None:
```

Comment:
I need to read more here to get my head around the new class, but my immediate reaction to methods that are version specific is that I'd hope they could live in the `Format` object.

Reply:
I'm trying to think this through... It's possible for everything to become version specific if it changes in a future version, and we wouldn't want the API to change when methods move to a `Format` object.
Everything in the `Axes` class is version specific, (e.g. `_validate_axes_types` will soon become more liberal). So do we move those internal methods to `Format` objects when they change?
This will mean that eventually all the code for everything will migrate `Format` objects.
Are the methods of `Format` objects (e.g. `validate_axes()`) expected to be part of the external API?
What's the use-case or benefit of all version-specific code being in `Format` objects?

Reply:
Trying to think of another example with the 0.4 HCS changes to the `wells` element discussed in https://github.com/ome/ome-zarr-py/pull/157, I could certainly imagine migrating `_validate_well_images(wells)` by `fmt.validate_well_images(wells)` or similar and let `FormatV04` vs `FormatV{01,02,03}` override the methods as appropriate.